lib/logstash/outputs/elasticsearch/http_client.rb

```diff
@@ -29,24 +29,45 @@ def template_install(name, template, force=false)
       template_put(name, template)
     end
 
-    def bulk(actions)
-      return if actions.empty?
-      bulk_body = actions.collect do |action, args, source|
-        if action == 'update'
-          if args[:_id]
-            source = { 'doc' => source }
-            if @options[:doc_as_upsert]
-              source['doc_as_upsert'] = true
-            else
-              source['upsert'] = args[:_upsert] if args[:_upsert]
-            end
+    def process_update_action(args, source)
+      if args[:_id]
+        if args[:_script]
```

Comment:
This if block is fairly large. Have you considered breaking it up into multiple functions? Specifically the main if clause is larger than we'd like for readability / testability.